### PR TITLE
ENH: Add seed option for gaussian_kde.resample

### LINKED ISSUE #10585: Use RandomState for gaussian_kde.resample
Currently, `scipy.stats.gaussian_kde.resample` use  `numpy.random.choice` and  `numpy.multivariate_normal`. However, no `RandomState` option is available for these two functions. It will be much better if one can use `RandomState` for the resampling method.

### PR DESCRIPTION
**Pull request by yqiuu**
#### Reference issue

 This PR aims to fix #10585.

#### What does this implement/fix?

An option `seed` is added to `gaussian_kde.resample`. The method now draws random variables using `numpy.random.RandomState`, which make the results reproducible.

### CODE REVIEW

**Review by tylerjereddy — CHANGES_REQUESTED** [review]
Would a simple unit test be possible for the new argument support?

**Comment by yqiuu** [comment]
Unit test added.

File: scipy/stats/kde.py [hunk 1]
```diff
@@ -447,6 +447,12 @@ def resample(self, size=None):
             The number of samples to draw.  If not provided, then the size is
             the same as the effective number of samples in the underlying
             dataset.
+        seed : None or int or ``numpy.random.RandomState`` instance, optional
```
Comment (andyfaff) [inline scipy/stats/kde.py]: Could you change the docstring to:
```
    seed : int or `np.random.RandomState`, optional
        If `seed` is not specified the `np.RandomState` singleton is used.
        If `seed` is an int, a new `np.random.RandomState` instance is used,
        seeded with seed.
        If `seed` is already a `np.random.RandomState instance`, then that
        `np.random.RandomState` instance is used.
        Specify `seed` for reproducible drawing of random variates.
```

File: scipy/stats/tests/test_kdeoth.py [hunk 2]
```diff
@@ -390,3 +390,47 @@ def test_weights_integer():
...
+def test_seed():
+    # Test the seed option of the resample method
+    def test_seed_sub(gkde_trail):
+        n_sample = 200
+        # The results should be different without using seed
+        samp1 = gkde_trail.resample(n_sample)
+        samp2 = gkde_trail.resample(n_sample)
+        assert_raises(
+            AssertionError, assert_array_almost_equal,
+            samp1, samp2, decimal = 13
```
Comment (andyfaff) [inline scipy/stats/tests/test_kdeoth.py]: PEP8, remove the spaces: `decimal=13`

Comment (andyfaff) [reply scipy/stats/tests/test_kdeoth.py]: In fact could you use `assert_allclose` here, with an `atol=1e-13`? If they're not supposed to be the same then perhaps the test tolerance can be relaxed to make the test harder?

File: scipy/stats/tests/test_kdeoth.py [hunk 3]
```diff
@@ -390,3 +390,47 @@ def test_weights_integer():
...
+        )
+        # Use integer seed
+        seed = 831
+        samp1 = gkde_trail.resample(n_sample, seed = seed)
```
Comment (andyfaff) [inline scipy/stats/tests/test_kdeoth.py]: PEP8, remove space around equals sign,  `seed=seed`

File: scipy/stats/tests/test_kdeoth.py [hunk 4]
```diff
@@ -390,3 +390,47 @@ def test_weights_integer():
...
+        samp2 = gkde_trail.resample(n_sample, seed = seed)
```
Comment (andyfaff) [inline scipy/stats/tests/test_kdeoth.py]: and here

File: scipy/stats/tests/test_kdeoth.py [hunk 5]
```diff
@@ -390,3 +390,47 @@ def test_weights_integer():
...
+        assert_array_almost_equal(samp1, samp2, decimal = 13)
```
Comment (andyfaff) [inline scipy/stats/tests/test_kdeoth.py]: `assert_allclose(samp1, samp2, atol=1e-13)`

File: scipy/stats/tests/test_kdeoth.py [hunk 6]
```diff
@@ -390,3 +390,47 @@ def test_weights_integer():
...
+        # Use RandomState
+        rstate1 = np.random.RandomState(seed = 138)
```
Comment (andyfaff) [inline scipy/stats/tests/test_kdeoth.py]: PEP8, remove space around equals sign

File: scipy/stats/tests/test_kdeoth.py [hunk 7]
```diff
@@ -390,3 +390,47 @@ def test_weights_integer():
...
+        samp1 = gkde_trail.resample(n_sample, seed = rstate1)
```
Comment (andyfaff) [inline scipy/stats/tests/test_kdeoth.py]: and here

File: scipy/stats/tests/test_kdeoth.py [hunk 8]
```diff
@@ -390,3 +390,47 @@ def test_weights_integer():
...
+        rstate2 = np.random.RandomState(seed = 138)
```
Comment (andyfaff) [inline scipy/stats/tests/test_kdeoth.py]: and here

File: scipy/stats/tests/test_kdeoth.py [hunk 9]
```diff
@@ -390,3 +390,47 @@ def test_weights_integer():
...
+        samp2 = gkde_trail.resample(n_sample, seed = rstate2)
```
Comment (andyfaff) [inline scipy/stats/tests/test_kdeoth.py]: and here

File: scipy/stats/tests/test_kdeoth.py [hunk 10]
```diff
@@ -390,3 +390,47 @@ def test_weights_integer():
...
+        assert_array_almost_equal(samp1, samp2, decimal = 13)
```
Comment (andyfaff) [inline scipy/stats/tests/test_kdeoth.py]: assert_allclose(samp1, samp2, atol=1e-13)

**Review by andyfaff — CHANGES_REQUESTED** [review]
I've made some minor comments, nothing serious. Mainly PEP8, and a request to change the docstring (to one that we would use elsewhere).
Otherwise looks good, and can merge once the changes are made.

File: scipy/stats/kde.py (hunk 1 above)
Comment (yqiuu) [reply scipy/stats/kde.py]: Should `np.RandomState` in the second line be `np.random.RandomState`?

Comment (andyfaff) [reply scipy/stats/kde.py]: Perhaps change to:
```
If `seed` is None, return the RandomState singleton used by np.random.
```

Comment (yqiuu) [reply scipy/stats/kde.py]: I think that this function does not return any `RandomState`. Thus, I have changed to:
``If `seed` is None, random variates are drawn by the RandomState singleton used by np.random.``
Let me know if any further changes are required.

**Comment by andyfaff** [comment]
Great work, thank you for contributing to scipy @yqiuu. Going to squash and merge this now. 


In future would it be possible to prefix your commit messages with one of the following, it helps those scanning the git history go through the lines a bit more quickly.

API: an (incompatible) API change
BENCH: changes to the benchmark suite
BLD: change related to building numpy
BUG: bug fix
DEP: deprecate something, or remove a deprecated object
DEV: development tool or utility
DOC: documentation
ENH: enhancement
MAINT: maintenance commit (refactoring, typos, etc.)
REV: revert an earlier commit
STY: style fix (whitespace, PEP8)
TST: addition or modification of tests
REL: related to releasing numpy

https://docs.scipy.org/doc/numpy-1.13.0/dev/gitwash/development_workflow.html

**Pull request merged**